EventGenerator/src/GammaConvGenerator_module.cc

```diff
@@ -0,0 +1,153 @@
+    , simsToken_{consumes<SimParticleCollection>(conf().inputSimParticles())}
+    , verbosity_{conf().verbosity()}
+    , eng_{createEngine(art::ServiceHandle<SeedService>()->getSeed())}
+    , randFlat_{eng_}
+  {
+    produces<mu2e::StageParticleCollection>();
+    process = ProcessCode::mu2eGammaConversion;
+  }
+
+  //================================================================
+  void GammaConvGenerator::produce(art::Event& event) {
+    auto output{std::make_unique<StageParticleCollection>()};
+
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_);
+    const auto gammas = simParticleList(simh,PDGCode::gamma,ProcessCode::conv);
+
+    if(gammas.empty()) {
+      throw   cet::exception("BADINPUT")
+        <<"GammaConvGenerator::produce(): no suitable converted gamma in the input SimParticleCollection\n";
+    }
+
+    // Take the converted gamma as SimParticle
+    const auto gammapart = gammas.at(eng_.operator unsigned int() % gammas.size());
+    const auto p4_gamma = gammapart->endMomentum();
+    const auto p_gamma   = p4_gamma.vect().mag();
+    const auto t_gammastop = gammapart->endGlobalTime();
+
+    // Start generating the e-:
+    // extract flat momentum between 0 and p_gamma
+    double p_eminus = randFlat_.fire(0., p_gamma);
+    double E_eminus = sqrt(particleMass_*particleMass_ + p_eminus*p_eminus);
+    // extract flat between 0 and 2pi the phi angle wrt gamma
+    double phi_eminus = randFlat_.fire(0., CLHEP::twopi);
+    // Fix to Mass_electron/Egamma radians the theta angle wrt to gamma
+    double theta_eminus = particleMass_/p_gamma;
+    // rotate so that the photon direction is the new z axis
+    const auto dir_gamma = p4_gamma.vect().unit();
+    double px_eminus = p_eminus * sin(theta_eminus) * cos (phi_eminus);
+    double py_eminus = p_eminus * sin(theta_eminus) * sin (phi_eminus);
+    double pz_eminus = p_eminus * cos(theta_eminus);
+    CLHEP::HepLorentzVector p4_eminus_temp(px_eminus, py_eminus, pz_eminus, E_eminus);
+    CLHEP::HepLorentzVector p4_eminus= p4_eminus_temp.rotateUz(dir_gamma);
+
+    output->emplace_back(gammapart,
+                         process,
+                         PDGCode::e_minus,
+                         gammapart->endPosition(),
+                         p4_eminus,
+                         t_gammastop
+                         );
+
+    // Now impose momentum conservation to find positron momentum
+
+    double px_eplus = -px_eminus;
+    double py_eplus = -py_eminus;
+    double pz_eplus = p_gamma-pz_eminus;
+    double p_eplus  = sqrt( px_eplus*px_eplus +
+                            py_eplus*py_eplus +
+                            pz_eplus*pz_eplus);
+    double E_eplus = sqrt(particleMass_*particleMass_ + p_eplus*p_eplus );
+    CLHEP::HepLorentzVector p4_eplus_temp(px_eplus, py_eplus, pz_eplus, E_eplus);
+    CLHEP::HepLorentzVector p4_eplus= p4_eplus_temp.rotateUz(dir_gamma);
+
+    output->emplace_back(gammapart,
```

Comment:
IIRC the mass of the e+e- pair follows a roughly exponential distribution, peaking at 2me.   What is the assumption here?  0 mass?  Is there any plan to provide a mass distribution in future?

Reply:
I impose momentum conservation neglecting nucleus recoil and the invariant mass is peaked at 2me I have added slide 9 to doc-db 47028 to show this